altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+        // divides it in two.
+        else {
+            unsafe {
+                (*current).block_size = current_size - using_size - align_diff;
+                let upper_block = self.shift_block_forward(current, using_size + align_diff);
+                ptr::write(current as *mut BlockHeader, BlockHeader::new(align_diff));
+                (*current).next_block = upper_block;
+            }
+        }
+        alloc_start as *mut u8
+    }
+
+    /*
+    TODO: We can still implement merging to deal with fragmentation
+    - Nothing adjacent: Make new block, connect to closest blocks
+    - Adjacent at tail: Merge with tail block, move block, switch leading ptr
+    - Adjacent at lead: Merge with lead, no additional changes
+    - Adjacent at both: Merge two with lead (add sizes, switch lead ptr)
+    */
+
+    // Deallocates memory, placing it back in the free list for later use
+    // Adds a free block to the list based on alloc_ptr so that the list remains
+    // sorted based on memory position.
+    pub fn deallocate(&mut self, alloc_ptr: *mut u8, size: usize, _align: usize) {
+        // We can immediately add the block at the deallocated position
+        let alloc_block_ptr = alloc_ptr as *mut BlockHeader;
+        let used_memory = alignment::align_up(size, self.block_hdr_size);
+        unsafe {
+            ptr::write(&mut *alloc_block_ptr, BlockHeader::new(used_memory));
+        }
+
+        // Memory location to be added at head of the list if the list is empty
+        // or if it's position in memory comes before the current head.
+        if self.head.is_null() || alloc_block_ptr < self.head {
+            unsafe { (*alloc_block_ptr).next_block = self.head; }
+            self.head = alloc_block_ptr;
+            return;
+        }
+        let (mut previous, current) = {
+            let free_mem_between = |previous, current| {
+                if alloc_block_ptr == current {
+                    panic!("attempt to free memory that's already free");
+                }
+                previous < alloc_block_ptr && alloc_block_ptr < current
+            };
+            self.find_block(&free_mem_between)
+        };
+        unsafe {
+            (*previous).next_block = alloc_block_ptr;
+            (*alloc_block_ptr).next_block = current;
+        }
+    }
+
+    // This relocates BlockHeaders in memory, used when we do allocations.
+    fn shift_block_forward(&self, current_pos: *mut BlockHeader, offset_val: usize) -> *mut BlockHeader {
+        // If we don't convert this, offset does not work correctly
+        let current_ptr = current_pos as *mut u8;
+        unsafe {
+            let new_pos = current_ptr.offset(offset_val as isize) as *mut BlockHeader;
+            let current_block = ptr::read(current_pos);
+            *new_pos = current_block;
+            new_pos as *mut BlockHeader
+        }
+    }
+}
+
+// For testing purposes, we make versions of the list structures with public data
```

Comment:
This was related to the comment I have further up about having two versions of this struct, this one for testing (because some functions in test.rs are using the fields). Should those be relocated back to this file to avoid having to do this? I put them on the TestFreeList struct because they seemed to make sense there.